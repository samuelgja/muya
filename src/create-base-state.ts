import type { Emitter } from './create-emitter'
import { merge } from './merge'
import { select } from './select'
import type { BaseState, GetterState } from './types'

interface Options<T> {
  readonly emitter: Emitter<T>
  readonly reset: () => void
  readonly getState: () => T
  readonly getGetterState: () => GetterState<T>
}
export function createBaseState<T>(options: Options<T>): BaseState<T> {
  const { emitter, getGetterState, reset, getState } = options
  return {
    getState,
    reset,
    select(selector, isSame) {
      const state = getGetterState()
      return select(state, selector, isSame)
    },
    merge(state2, selector, isEqualHook) {
      const state = getGetterState()
      return merge(state, state2, selector, isEqualHook)
    },
    __internal: {
      emitter,
    },
    subscribe(listener) {
      listener(getState())
      return emitter.subscribe(() => {
        listener(getState())
      })
    },
  }
}
