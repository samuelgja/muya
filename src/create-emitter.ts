export type EmitterSubscribe<P = undefined> = (listener: (...params: P[]) => void) => () => void
export interface Emitter<T, R = T, P = undefined> {
  subscribe: EmitterSubscribe<P>
  getSnapshot: () => R
  emit: (...params: P[]) => void
  size: number
  clear: () => void
}

export function createEmitter<T, R = T, P = undefined>(getSnapshot: () => R): Emitter<T, R, P> {
  const listeners = new Set<(...params: P[]) => void>()
  return {
    clear: () => {
      listeners.clear()
    },
    subscribe: (listener) => {
      listeners.add(listener)
      return () => {
        listeners.delete(listener)
      }
    },
    emit: (...params) => {
      for (const listener of listeners) {
        listener(...params)
      }
    },
    getSnapshot,
    size: listeners.size,
  }
}
