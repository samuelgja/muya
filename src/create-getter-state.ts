import type { BaseState, GetterState } from './types'
import { useStateValue } from './use-state-value'

interface Options<T> {
  readonly baseState: BaseState<T>
}
export function createGetterState<T>(options: Options<T>): GetterState<T> {
  const { baseState } = options
  const useSliceState: GetterState<T> = (useSelector, isEqualHook) => {
    return useStateValue(useSliceState, useSelector, isEqualHook)
  }
  useSliceState.__internal = baseState.__internal
  useSliceState.getState = baseState.getState
  useSliceState.reset = baseState.reset
  useSliceState.select = baseState.select
  useSliceState.merge = baseState.merge
  useSliceState.subscribe = baseState.subscribe
  return useSliceState
}
