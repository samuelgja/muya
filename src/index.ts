export * from './types'
export { create } from './create'
export { useCreate } from './use-state-value'
export { shallow } from './shallow'
