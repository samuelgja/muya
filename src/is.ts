import type { Ref, Setter, SetValue } from './types'

export function isPromise(value: unknown): value is Promise<unknown> {
  return value instanceof Promise
}
export function isFunction(value: unknown): value is (...args: unknown[]) => unknown {
  return typeof value === 'function'
}
export function isSetValueFunction<T>(value: SetValue<T>): value is Setter<T> {
  return typeof value === 'function'
}
export function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}
export function isRef<T>(value: unknown): value is Ref<T> {
  return isObject(value) && value.isRef === true
}

export function isMap(value: unknown): value is Map<unknown, unknown> {
  return value instanceof Map
}

export function isSet(value: unknown): value is Set<unknown> {
  return value instanceof Set
}

export function isArray(value: unknown): value is Array<unknown> {
  return Array.isArray(value)
}

export function isEqualBase<T>(valueA: T, valueB: T): boolean {
  if (valueA === valueB) {
    return true
  }
  return !!Object.is(valueA, valueB)
}
