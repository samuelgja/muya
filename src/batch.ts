import { isAbortError, isAsyncFunction, isPromise, isSetValueFunction } from './is'
import type { PromiseAndValue, SetValue } from './types'

interface Options<T> {
  setValue: (value: T) => void
  getValue: () => T

  onFlush: (current: Awaited<T>) => void
}
interface BatchResult<T> {
  current?: T
  batches: SetValue<T>[]
  addValue: (value: SetValue<T>) => void
  flush: () => void
  abortController?: AbortController
}
export function createBatcher<T>(options: Options<T>) {
  const batches: SetValue<T>[] = []

  function setState(value: SetValue<T>) {
    if (batch.current == undefined) {
      batch.current = options.getValue()
    }
    if (!isSetValueFunction(value)) {
      if (batch.abortController) {
        batch.abortController.abort()
      }
      batch.current = value
      return
    }

    const result = value(batch.current as PromiseAndValue<T>)
    if (!isPromise(result)) {
      batch.current = result as T
      return
    }

    return result
      .then((resolvedValue) => {
        // check if state.value is resolved value
        if (isPromise(batch.current) && batch.abortController) {
          batch.abortController.abort()
        }
        batch.current = resolvedValue as T
      })
      .catch((error) => {
        if (isAbortError(error)) {
          return
        }
      })
  }

  function scheduler() {
    const frameSize = batches.length
    queueMicrotask(() => {
      const frameSizeDiff = batches.length - frameSize
      if (frameSizeDiff < 1 && batch.batches.length > 0) {
        batch.flush()
      }
    })
    // requestAnimationFrame(() => {
    //   const frameSizeDiff = batches.length - frameSize
    //   console.log({ frameSizeDiff })
    //   if (frameSizeDiff < 1 && batch.batches.length > 0) {
    //     batch.flush()
    //   }
    // })
  }

  async function flush() {
    for (const batchItem of batch.batches) {
      if (isAsyncFunction(batchItem)) {
        // eslint-disable-next-line sonarjs/no-invalid-await
        await setState(await batchItem)
      } else {
        setState(batchItem)
      }
    }
    batch.batches = []
    options.onFlush(batch.current as Awaited<T>)
  }

  function addValue(value: SetValue<T>) {
    batch.batches.push(value)
    scheduler()
  }
  const batch: BatchResult<T> = {
    current: undefined,
    batches,
    addValue,
    flush,
  }
  return batch
}
