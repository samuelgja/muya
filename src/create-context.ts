import { isPromise } from './is'

const EMPTY_CONTEXT = Symbol('_')

/**
 * Base context interface.
 */
export function createContext<T>(defaultContextValue: T) {
  // Initialize the context with the default value
  let currentContext: T | typeof EMPTY_CONTEXT = defaultContextValue ?? EMPTY_CONTEXT

  /**
   * Retrieves the current context value.
   */
  function use(): T {
    return currentContext === EMPTY_CONTEXT ? defaultContextValue : currentContext
  }

  /**
   * Runs a callback within a new context.
   * @param ctxValue The new context value.
   * @param cb The callback to execute.
   */
  function run<R>(ctxValue: T, cb: () => R): R | Promise<R> {
    const previousContext = currentContext
    currentContext = ctxValue

    const result = cb()
    const isResultPromise = isPromise(result)
    try {
      if (isResultPromise) {
        // If the callback returns a promise, ensure context is reset after it resolves
        return (async () => {
          try {
            return await result
          } finally {
            currentContext = previousContext
          }
        })()
      }
      // For synchronous callbacks, reset context immediately
      return result
    } finally {
      // Reset context after the callback completes
      if (!isResultPromise) {
        currentContext = previousContext
      }
    }
  }

  /**
   * Wraps an asynchronous callback to preserve the current context.
   * @param cb The asynchronous callback to wrap.
   */
  function wrap<X>(cb: () => X | Promise<X>): () => X | Promise<X> {
    const capturedContext = use()
    return async () => {
      const previousContext = currentContext
      currentContext = capturedContext
      try {
        return await cb()
      } finally {
        currentContext = previousContext
      }
    }
  }

  return {
    run,
    use,
    wrap, // Function to wrap async callbacks
  }
}
