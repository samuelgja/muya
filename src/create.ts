import { createContext } from './create-context'
import type { Emitter } from './create-emitter'
import { createEmitter } from './create-emitter'
import { isAbortError, isFunction, isPromise } from './is'
import { cancelablePromise } from './common'
import type { IsEqual, Set, SetValue } from './types'
import { createBatcher } from './batch'

const context = createContext<StateNotGeneric | undefined>(undefined)

type DefaultValue<T> = T | (() => T)

interface StateNotGeneric {
  get: () => unknown
  reset: () => void
  value?: unknown
  id: string
  subscribe: (listener: (value: unknown) => void) => () => void
  remove: () => void
}
export interface GetState<T> extends StateNotGeneric {
  <S>(selector?: (stateValue: T) => S, isEqual?: IsEqual<S>): undefined extends S ? T : S
  emitter: Emitter<T>
  get: () => T
  reset: () => void
  value?: T
  id: string
  subscribe: (listener: (value: T) => void) => () => void
}
interface State<T> extends GetState<T> {
  set: Set<T>
}

// it still can be promise, this just handle if defaultValue is function
export function getDefaultValue<T>(state: State<T>, defaultValue: DefaultValue<T>): T {
  const value = isFunction(defaultValue) ? context.run(state, defaultValue) : defaultValue
  return value as T
}
let id = 0
function getId() {
  id++
  return id.toString(36)
}

export function create<T extends () => T>(defaultValue: DefaultValue<T>): GetState<ReturnType<T>>
export function create<T>(defaultValue: DefaultValue<T>): State<T>
export function create<T>(defaultValue: DefaultValue<T>): GetState<T> | State<T> {
  const contextSubscriptions = new Map<string, () => void>()
  let previousData: unknown
  const state: State<T> = <S>(
    selector: (stateValue: T) => S = (stateValue) => stateValue as unknown as S,
    isEqual: IsEqual<S> = (a, b) => a === b,
  ): undefined extends S ? T : S => {
    const ctx = context.use()
    const hasCtx = !!ctx
    if (!hasCtx) {
      throw new Error(`calling states must be only inside of a running context.`)
    }

    if (!contextSubscriptions.has(ctx.id)) {
      const unsubscribe = state.emitter.subscribe(() => {
        console.log('Calling reset from ', state.id)
        console.log('For derive context: ', ctx.id)
        ctx.reset()
      })
      contextSubscriptions.set(ctx.id, unsubscribe)
    }
    const selectedValue = selector(getValue())

    if (previousData !== undefined && isEqual(previousData as S, selectedValue)) {
      return previousData as undefined extends S ? T : S
    }
    previousData = selectedValue as undefined

    return selectedValue as undefined extends S ? T : S
  }

  const batch = createBatcher<T>({
    getValue,
    setValue: setValueRaw,
    onFlush: (current) => {
      console.log('flush', state.id, current)
      state.value = current
      state.emitter.emit()
    },
  })

  state.emitter = createEmitter(getValue)

  let isResolving = false
  function resolveValue(): T {
    if (isResolving) {
      return state.value as T
    }
    const value = getDefaultValue(state, defaultValue)
    if (!isPromise<T>(value)) {
      return value
    }

    isResolving = true
    const { promise, controller } = cancelablePromise<T>(value, batch.abortController)
    batch.abortController = controller
    promise
      .then((resolvedValue) => {
        isResolving = false
        state.value = resolvedValue
        state.emitter.emit()
      })
      .catch((error) => {
        isResolving = false

        if (isAbortError(error)) {
          return
        }
      })
    // switch Promise to cancelable promise
    return promise as unknown as T
  }

  function getValue() {
    if (state.value === undefined) {
      state.value = resolveValue()
    }
    return state.value
  }

  state.reset = function call() {
    state.value = resolveValue()
    state.emitter.emit()
  }

  state.get = getValue
  state.id = getId()
  state.remove = function remove() {
    for (const [, unsubscribe] of contextSubscriptions) {
      unsubscribe()
    }
    contextSubscriptions.clear()
    state.emitter.clear()
  }

  if (isFunction(defaultValue)) {
    return state
  }
  state.subscribe = function subscribe(listener: (value: T) => void) {
    return state.emitter.subscribe(() => {
      listener(state.get())
    })
  }

  // function setState(value: SetValue<T>) {
  //   if (state.value === undefined) {
  //     state.value = resolveValue()
  //   }

  //   if (!isSetValueFunction(value)) {
  //     if (batch.abortController) {
  //       batch.abortController.abort()
  //     }

  //     state.value = value
  //     state.emitter.emit()
  //     return
  //   }

  //   const result = value(state.value as PromiseAndValue<T>)
  //   if (!isPromise(result)) {
  //     state.value = result as T
  //     state.emitter.emit()
  //     return
  //   }
  //   result
  //     .then((resolvedValue) => {
  //       // check if state.value is resolved value
  //       if (isPromise(state.value) && batch.abortController) {
  //         batch.abortController.abort()
  //       }
  //       state.value = resolvedValue as T
  //       state.emitter.emit()
  //     })
  //     .catch((error) => {
  //       if (isAbortError(error)) {
  //         return
  //       }
  //     })
  // }

  function setValueRaw(value: T) {
    state.value = value
    state.emitter.emit()
  }

  function setStateBatch(value: SetValue<T>) {
    // setState(value)
    // console.log(value)
    batch.addValue(value)
  }

  state.set = setStateBatch

  return state
}
