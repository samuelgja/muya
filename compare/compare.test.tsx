import { renderHook, act } from '@testing-library/react-hooks'
import { waitFor } from '@testing-library/react'
import { atom, useAtom, useAtomValue } from 'jotai'
import { create, use } from '../src'
import { Suspense, useLayoutEffect } from 'react'
import { create as zustandCreate } from 'zustand'

describe('should count re-renders', () => {
  const jotaiReRendersBefore = jest.fn()
  const jotaiReRendersAfter = jest.fn()
  const muyaReRendersBefore = jest.fn()
  const muyaReRendersAfter = jest.fn()
  beforeEach(() => {
    jotaiReRendersAfter.mockClear()
    jotaiReRendersBefore.mockClear()
    muyaReRendersAfter.mockClear()
    muyaReRendersBefore.mockClear()
  })
  it('should compare counter', () => {
    const jotaiCounter = atom(0)

    const muyaCounter = create(0)

    const startTime = Date.now()
    const jotaiResult = renderHook(() => {
      jotaiReRendersBefore()
      const [counter] = useAtom(jotaiCounter)
      jotaiReRendersAfter()
      return counter
    })
    const endTime = Date.now()
    console.log('Jotai render time:', endTime - startTime)
    expect(jotaiResult.result.current).toBe(0)

    // JOTAI RE_RENDER 2x TIMES
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(2)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(2)

    const startTimeMuya = Date.now()
    const muyaResult = renderHook(() => {
      muyaReRendersBefore()
      const counter = use(muyaCounter)
      muyaReRendersAfter()
      return counter
    })
    const endTimeMuya = Date.now()
    console.log('Muya render time:', endTimeMuya - startTimeMuya)
    expect(muyaResult.result.current).toBe(0)
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(1)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(1)

    // CREATE SNAPSHOTS
    expect(jotaiResult.result.current).toMatchSnapshot("Jotai's counter")
    expect(jotaiReRendersBefore.mock.calls).toMatchSnapshot("Jotai's re-renders")
    expect(jotaiReRendersAfter.mock.calls).toMatchSnapshot("Jotai's re-renders")

    expect(muyaResult.result.current).toMatchSnapshot("Muya's counter")
    expect(muyaReRendersBefore.mock.calls).toMatchSnapshot("Muya's re-renders")
    expect(muyaReRendersAfter.mock.calls).toMatchSnapshot("Muya's re-renders")

    // UPDATE JOTAI COUNTER
  })
  it('should compare counter with set', async () => {
    const jotaiCounter = atom(0)

    const muyaCounter = create(0)

    const jotaiResult = renderHook(() => {
      jotaiReRendersBefore()
      const [counter, setCounter] = useAtom(jotaiCounter)
      jotaiReRendersAfter()
      useLayoutEffect(() => {
        setCounter((c) => c + 1)
      }, [])
      return { counter, setCounter }
    })

    const muyaResult = renderHook(() => {
      muyaReRendersBefore()
      const counter = use(muyaCounter)
      muyaReRendersAfter()
      useLayoutEffect(() => {
        muyaCounter.set((c) => c + 1)
      }, [])
      return counter
    })

    expect(jotaiResult.result.current.counter).toBe(1)
    expect(muyaResult.result.current).toBe(1)

    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(2)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(2)

    expect(muyaReRendersBefore).toHaveBeenCalledTimes(2)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(2)
  })

  it('should compare counter with set and async', () => {
    const jotaiText = atom('readonly atoms')
    const jotaiUppercase = atom((get) => get(jotaiText).toUpperCase())

    const muyaText = create('readonly atoms')
    const muyaUppercase = create(() => muyaText().toUpperCase())

    const jotaiResult = renderHook(() => {
      jotaiReRendersBefore()
      const text = useAtom(jotaiText)
      const uppercase = useAtom(jotaiUppercase)
      jotaiReRendersAfter()
      return { text, uppercase }
    })

    const muyaResult = renderHook(() => {
      muyaReRendersBefore()
      const text = use(muyaText)
      const uppercase = use(muyaUppercase)
      muyaReRendersAfter()
      return { text, uppercase }
    })

    expect(jotaiResult.result.current.text[0]).toBe('readonly atoms')
    expect(jotaiResult.result.current.uppercase[0]).toBe('READONLY ATOMS')
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(2)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(2)

    expect(muyaResult.result.current.text).toBe('readonly atoms')
    expect(muyaResult.result.current.uppercase).toBe('READONLY ATOMS')
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(1)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(1)

    act(() => {
      jotaiResult.result.current.text[1]('readonly atoms2')
      muyaText.set('readonly atoms2')
    })

    expect(jotaiResult.result.current.text[0]).toBe('readonly atoms2')
    expect(jotaiResult.result.current.uppercase[0]).toBe('READONLY ATOMS2')
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(3)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(3)

    expect(muyaResult.result.current.text).toBe('readonly atoms2')
    expect(muyaResult.result.current.uppercase).toBe('READONLY ATOMS2')
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(2)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(2)
  })

  it('should compare counter with set and async', async () => {
    const jotaiCounter = atom(1)
    const counterAsync = atom(async (get) => get(jotaiCounter) * 5)

    const muyaCounter = create(1)
    const counterAsyncMuya = create(async () => muyaCounter() * 5)

    const jotaiResult = renderHook(
      () => {
        jotaiReRendersBefore()
        const counter = useAtom(jotaiCounter)
        const asyncCounter = useAtomValue(counterAsync)
        jotaiReRendersAfter()
        return { counter, asyncCounter }
      },
      { wrapper: ({ children }) => <Suspense fallback="loading">{children}</Suspense> },
    )

    await waitFor(() => {
      expect(jotaiResult.result.current.counter[0]).toBe(1)
      expect(jotaiResult.result.current.asyncCounter).toBe(5)
    })

    expect(jotaiResult.result.current.counter[0]).toBe(1)
    expect(jotaiResult.result.current.asyncCounter).toBe(5)
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(3)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(2)

    const muyaResult = renderHook(
      () => {
        muyaReRendersBefore()
        const counter = use(muyaCounter)
        const asyncCounter = use(counterAsyncMuya)
        muyaReRendersAfter()
        return { counter, asyncCounter }
      },
      { wrapper: ({ children }) => <Suspense fallback="loading">{children}</Suspense> },
    )

    await waitFor(() => {
      expect(muyaResult.result.current.counter).toBe(1)
      expect(muyaResult.result.current.asyncCounter).toBe(5)
    })

    expect(muyaResult.result.current.counter).toBe(1)
    expect(muyaResult.result.current.asyncCounter).toBe(5)
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(2)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(1)

    act(() => {
      jotaiResult.result.current.counter[1](2)
      muyaCounter.set(2)
    })

    await waitFor(() => {
      expect(jotaiResult.result.current.counter[0]).toBe(2)
      expect(jotaiResult.result.current.asyncCounter).toBe(10)
    })

    expect(jotaiResult.result.current.counter[0]).toBe(2)
    expect(jotaiResult.result.current.asyncCounter).toBe(10)
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(5)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(3)

    expect(muyaResult.result.current.counter).toBe(2)
    expect(muyaResult.result.current.asyncCounter).toBe(10)
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(4)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(2)
  })
  it('should compare counter with set and async with fetch', async () => {
    const counter = atom(1)
    const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))
    const asyncAtom = atom(async (get) => {
      await sleep(100)
      return get(counter) + 1
    })

    const muyaCounter = create(1)
    const asyncAtomMuya = create(async () => {
      await sleep(100)
      return muyaCounter() + 1
    })

    const jotaiResult = renderHook(
      () => {
        jotaiReRendersBefore()
        const counterValue = useAtom(counter)
        const asyncValue = useAtomValue(asyncAtom)
        jotaiReRendersAfter()
        return { counterValue, asyncValue }
      },
      { wrapper: ({ children }) => <Suspense fallback="loading">{children}</Suspense> },
    )

    await waitFor(() => {
      expect(jotaiResult.result.current.asyncValue).toBe(2)
    })

    expect(jotaiResult.result.current.counterValue[0]).toBe(1)
    expect(jotaiResult.result.current.asyncValue).toBe(2)
    expect(jotaiReRendersBefore).toHaveBeenCalledTimes(3)
    expect(jotaiReRendersAfter).toHaveBeenCalledTimes(2)

    const muyaResult = renderHook(
      () => {
        muyaReRendersBefore()
        const counterValue = use(muyaCounter)
        const asyncValue = use(asyncAtomMuya)
        muyaReRendersAfter()
        return { counterValue, asyncValue }
      },
      { wrapper: ({ children }) => <Suspense fallback="loading">{children}</Suspense> },
    )

    await waitFor(() => {
      expect(muyaResult.result.current.asyncValue).toBe(2)
    })

    expect(muyaResult.result.current.counterValue).toBe(1)
    expect(muyaResult.result.current.asyncValue).toBe(2)
    expect(muyaReRendersBefore).toHaveBeenCalledTimes(2)
    expect(muyaReRendersAfter).toHaveBeenCalledTimes(1)
  })

  it('should measure render times in loop of N count', async () => {
    const jotaiCounter = atom(0)
    const jotaiSum = atom((get) => get(jotaiCounter) + 1)
    const muyaCounter = create(0)
    const muyaSum = create(() => muyaCounter() + 1)

    const zustandCounter = zustandCreate((set) => ({
      bears: 0,
      increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
      removeAllBears: () => set({ bears: 0 }),
    }))
    const count = 2000

    const startTimeMuya = performance.now()
    for (let i = 0; i < count; i++) {
      const result = renderHook(() => {
        use(muyaSum)
        return use(muyaCounter)
      })
      act(() => {
        muyaCounter.set((c) => c + 1)
      })
    }
    const resultMuya = renderHook(() => {
      use(muyaSum)
      return use(muyaCounter)
    })
    await waitFor(() => {
      expect(resultMuya.result.current).toBe(count)
    })

    const endTimeMuya = performance.now()
    const totalTimeMuya = endTimeMuya - startTimeMuya

    const startTime = performance.now()
    for (let i = 0; i < count; i++) {
      const result = renderHook(() => {
        useAtom(jotaiSum)
        return useAtom(jotaiCounter)
      })
      act(() => {
        result.result.current[1]((c) => c + 1)
      })
    }
    const result = renderHook(() => {
      useAtom(jotaiSum)
      return useAtom(jotaiCounter)
    })
    await waitFor(() => {
      expect(result.result.current[0]).toBe(count)
    })
    const endTime = performance.now()
    const totalTime = endTime - startTime

    const startTimeZustand = performance.now()
    for (let i = 0; i < count; i++) {
      const result = renderHook(() => {
        const counter = zustandCounter()
        return counter
      })
      act(() => {
        result.result.current.increasePopulation()
      })
    }
    const resultZustand = renderHook(() => {
      const counter = zustandCounter()
      return counter
    })
    await waitFor(() => {
      expect(resultZustand.result.current.bears).toBe(count)
    })

    const endTimeZustand = performance.now()
    const totalTimeZustand = endTimeZustand - startTimeZustand

    console.log('Zustand render time:', totalTimeZustand)
    console.log('Jotai render time:', totalTime)
    console.log('Muya render time:', totalTimeMuya)

    expect(totalTimeMuya).toBeLessThan(totalTime)
    expect(totalTimeMuya).toBeLessThan(totalTimeZustand)
  })
})
