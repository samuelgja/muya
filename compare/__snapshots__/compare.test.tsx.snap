// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should count re-renders should measure render times in loop of N count: Jotai render time for 300 loops 1`] = `473`;

exports[`should count re-renders should measure render times in loop of N count: Muya render time for 300 loops 1`] = `505`;
